Add IPv6 availability check to skip tests when unavailable

diego-ciciani01 opened the pull request:
Skip IPv6 tests automatically when IPv6 is not available.

This fixes the problem that tests fail when IPv6 is not available on the system, which can worry users when they run `make test`.

IPv6 availibility is detected by opening a dummy server socket and trying to connect to it using a client socket over IPv6.

Fixes #2643

diego-ciciani01 committed:
Add IPv6 availability check to skip tests when unavailable

- Add is_ipv6_available() to detect IPv6 support

 - Add VALLKEY_REQUIRE_IPV6 env var (inside a 'if' condition) to force  IPv6 tests in CI, implement in the pipeline the export and volorization of env var

- Skip IPv6 tests automatically when IPv6 is not available

- BUG #2643

Signed-off-by: diego-ciciani01 <diego.ciciani@gmail.com>

zuiderkwast commented on tests/support/util.tcl:
```diff
@@ -499,6 +499,48 @@ proc roundFloat f {
+#2) netstat -an and filter on '*::1*'
+#3) in the and if both of tham fail, the last is the socket connection on ::1 address
+proc is_ipv6_available {} {
+    if {[catch {exec ifconfig lo0} result] == 0} {
+        if {[string match "*inet6*::1*" $result]} {
+            return 1
+        }
+    }
+    
+    if {[catch {exec netstat -an} result] == 0} {
+        if {[string match "*::1*" $result]} {
+            return 1
+        }
+    }
+    
+    if {[catch {
+        set sock [socket -async ::1 22]
+        close $sock
+        return 1
+    }]} {
+        return 0
+    }
```
Why do we need three ways? Can we use only the 3rd way to detect this and save the result in some global variable to avoid doing it multiple times?

diego-ciciani01 replied:
I agree with you, multiple checks can seem a bit repetitive, but during development, the third test on my macbook never worked (I'll leave a test I did in the comments, but I didn't add it to the PR). I also made sure that IPv6 was enabled, so as a workaround I also did the first two checks.

I currently have version 9 of tcl. 

here the the did

`proc is_ipv6_available {} {
    if {[catch {
        set sock [socket -async ::1 0]
        close $sock
    } err]} {
        puts "ipv6 check failed: $err"
        return 0
    }
    return 1 
}`

and i leave the result of the running test

<img width="1004" height="62" alt="Screenshot 2025-10-07 alle 11 43 23" src="https://github.com/user-attachments/assets/4f0ff918-7ba7-4b28-bd04-221fb85698ea" />

zuiderkwast replied:
Ah, you're testing to connect to port 22 = the SSH port. It is not always opened. We can not even assume that SSH is installed.

I think we can't assume that any specific port is opened. What we need to do is to create a server socket on a free port chosen by the OS. To do this, use port 0, a command like `socket -server ::1 0`. Read details under SERVER SOCKETS on the page https://www.tcl-lang.org/man/tcl/TclCmd/socket.html#M9

This is what I get on my system where IPv6 is available:

```
$ tclsh
% set s [socket -server ::1 0]
sockaaab2529cdd0
% chan configure $s -sockname
0.0.0.0 0.0.0.0 33953 :: :: 33953
```

Do the tests work with TCL 9? Do you get some error when running all the tests?

From what I remember, TCL 9 didn't work and I have opened #1673 to fix that.

zuiderkwast replied:
Try this:

```suggestion
    if {[catch {
        set sock [socket -server ::1 0]
        close $sock
        return 1
    }]} {
        return 0
    }
```

zuiderkwast replied:
I don't know how TCL behaves, but if `set sock [socket -server ::1 0]` never fails, even if IPv6 is not available, then we maybe also need to try to connect as a client to this server socket.

Something like this:

```tcl
    if {[catch {
        set server [socket -server ::1 0]
        set port [lindex [chan configure $server -sockname] 5]
        set client [socket ::1 $port]
        close $server
        close $client
        return 1
    }]} {
        return 0
    }
```

diego-ciciani01 replied:
I solved the problem. As you said, version 9 of TCL doesn't work very well. I switched to 8.6, and now the socket opens correctly. 

I will test also the client - server  connect

diego-ciciani01 replied:
why did you set set `port [lindex [chan configure $server -sockname] 5]` ? 

i thinks that the correct value should be 2, because this position returns the port

zuiderkwast replied:
I didn't find the docs for this but I tested locally in `tclsh` and I got the port on two places:

```
% chan configure $sock -sockname           
0.0.0.0 0.0.0.0 35783 :: :: 35783
```

It looks like IPv4 IPv4 port IPv6 IPv6 port. The last port maybe is for IPv6. That's what I was thinking. I don't know if it's correct. It's better to find the proper documentation for this.

diego-ciciani01 replied:
aah it make sense, so i found this  doc ->https://core.tcl-lang.org/tips/doc/trunk/tip/162.md
which explains exactly your test.

So i also added few commits with the changes that you told me

diego-ciciani01 replied:
aah it make sense, so i found this  doc ->https://core.tcl-lang.org/tips/doc/trunk/tip/162.md
which explains exactly your test.

So i also added few commits with the changes that you told me

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -225,18 +225,20 @@ start_server {tags {"introspection"}} {
-            set c [valkey ::1 [srv 0 port] 0 $::tls]
-            $c client setname "client-ipv6"
+    if {![should_skip_ipv6_tests]} {
+        start_server {tags {"ipv6"} overrides {bind {127.0.0.1 ::1}}} {
```
Let's use the existing tagging, here `tags {"ipv6"}`, that already exists and can be used for skipping tests. We can avoid changing the code in the individual test suites.

diego-ciciani01 replied:
You're absolutely right, I hadn't noticed the tag

zuiderkwast commented on tests/support/util.tcl:
```diff
@@ -499,6 +499,48 @@ proc roundFloat f {
+
+#Check the presence of env variable VALKEY_REQUIRE_IPV6 to decide if run the tests or not
+proc should_skip_ipv6_tests {} {
+    if {[info exists ::env(VALKEY_REQUIRE_IPV6)] && $::env(VALKEY_REQUIRE_IPV6) == "1"} {
```
Rather than an environment variable, we use flags like `--require-ipv6` or similar.

diego-ciciani01 replied:
yess, no problem

zuiderkwast reviewed (commented):
Good job. However, I think we should do it using the existing mechanism to skip tests based on the existing tagging.

The `tests/test_helper.tcl` is the entry point that checks command line arguments. Here, we can add "ipv6" to `::denytags` if IPv6 is not available and not required.

diego-ciciani01 commented:
hi @zuiderkwast, thank you very much for your feedback. If you need any further clarification, please let me know. Your advice is clear to me, please also let me know what you think about the IPv6 socket issue, after which I will modify the PR with your suggestions.

diego-ciciani01 committed:
feat: Refactor test utility and implement IPv6 feature tagging

Moves utility functions previously defined in  to the central test helper file, , for better organization and consistency across the testing framework.

Additionally, this commit introduces a new tag-based logic to conditionally run tests based on the availability of IPv6 support on the system. Tests requiring IPv6 connectivity are now tagged accordingly, ensuring that they are only executed when the necessary network configuration is present.

Signed-off-by: diego-ciciani01 <diego.ciciani@gmail.com>

diego-ciciani01 committed:
feat: Refactor test utility and implement IPv6 feature tagging

Moves utility functions previously defined in  to the central test helper file, , for better organization and consistency across the testing framework.

Additionally, this commit introduces a new tag-based logic to conditionally run tests based on the availability of IPv6 support on the system. Tests requiring IPv6 connectivity are now tagged accordingly, ensuring that they are only executed when the necessary network configuration is present.

Signed-off-by: diego-ciciani01 <diego.ciciani@gmail.com>

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -225,6 +225,7 @@ start_server {tags {"introspection"}} {
         assert_match *client-ip* $filtered
     } {}
 
+    
```
Don't add these extra blank lines. Keep the diff small and avoid this kind of whitespace changes.

```suggestion
```

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -240,6 +241,7 @@ start_server {tags {"introspection"}} {
         }
     }
 
+
```
```suggestion
```

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -264,8 +266,9 @@ start_server {tags {"introspection"}} {
         assert_error "*I/O error*" {$c1 ping}
     } {}
 
+    
```
```suggestion
```

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -279,6 +282,7 @@ start_server {tags {"introspection"}} {
             $c close
         }
     }
+    
```
```suggestion
```

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -509,14 +513,18 @@ start_server {tags {"introspection"}} {
         r client list not-ip $not_ip not-ip $not_ip
     } {}
-
+   
```
```suggestion

```

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -509,14 +513,18 @@ start_server {tags {"introspection"}} {
+
             regexp {addr=\[([a-fA-F0-9:]+)\]:\d+} $client_info -> ipv6only
             set filtered [$c client list not-ip "1.2.3.4" not-ip $ipv6only]
             assert_no_match *client-ipv6* $filtered
```
What's the purpose of adding this additional check?

It is very similar to teh check below, but is there a risk that matching either "1.2.3.4" or the client's IPv6 address will accidentally match because the client's IPv4 address happens to be 1.2.3.4?

diego-ciciani01 replied:
The duplicated lines probably appeared after I merged or rebased this branch with main while the PR was already open. 
I will remove this typo

zuiderkwast commented on tests/test_helper.tcl:
```diff
@@ -987,6 +990,29 @@ proc close_replication_stream {s} {
 }
 
+#IPv6 detection utilities
+#for this detaction: the socket connection on ::1 address
```
Add a space between the `#` and the text, for better readability.

```suggestion
# IPv6 detection utilities
# for this detaction: the socket connection on ::1 address
```

zuiderkwast commented on tests/test_helper.tcl:
```diff
@@ -987,6 +990,29 @@ proc close_replication_stream {s} {
+proc is_ipv6_available {} {
+        if {[catch {
+            set server [socket -server ::1 0]
+            set port [lindex [chan configure $server -sockname] 5]
```
I'm reading the documentation (https://www.tcl-lang.org/man/tcl/TclCmd/socket.html) for server sockets again. The server sockets, the syntax is

    socket -server command ?options? port

The IP address is not specified here. The `command` is the code to run when a client connects to it. The documentation says:

> For each connection Tcl will create a new channel that may be used to communicate with the client. Tcl then invokes command (properly a command prefix list, see the [EXAMPLES](https://www.tcl-lang.org/man/tcl/TclCmd/socket.html#M20) below) with three additional arguments: the name of the new channel, the address, in network address notation, of the client's host, and the client's port number.

Anyway, it seems to be OK to not have a correct command here, but I guess we should not write `::1` here because it's not an IP address at all. Better just use `dummy` or as the command (which will never be called anyway if the server doesn't start an event loop using vwait).

Also, if we use the option `-myaddr ::1`, then the server will only listen on the IPv6 address and the `chan configure` will only return 3 elements, where the last one is the port. That's also documented here under [CONFIGURATION OPTIONS, -sockname](https://www.tcl-lang.org/man/tcl/TclCmd/socket.html#M15):

> For server sockets this option returns a list of a multiple of three elements each group of which have the same meaning as described above. The list contains more than one group when the server socket was created without -myaddr or with the argument to -myaddr being a domain name that resolves multiple IP addresses that are local to the invoking host.

One more thing, the indentation here looks odd. Use 4 spaces for each level.

To sum up:

```suggestion
    if {[catch {
        set server [socket -server dummy -myaddr ::1 0]
        set port [lindex [chan configure $server -sockname] 2]
```

zuiderkwast commented on tests/test_helper.tcl:
```diff
@@ -987,6 +990,29 @@ proc close_replication_stream {s} {
+if {!$::require_ipv6 && ![is_ipv6_available]} {
+    lappend ::denytags "ipv6"
+    puts "IPv6 not available on this system, skipping IPv6 tests"
+}
```
I can see this will work, but comparing with other similar mechanisms to skip tests in various conditions, this logic should better be placed in the function `tags_acceptable` in the file `tests/support/server.tcl`. Add it next to these checks, here: https://github.com/valkey-io/valkey/blob/9.0.0-rc3/tests/support/server.tcl#L266

zuiderkwast reviewed (commented):
Great! This is getting closer. I have some more ideas though after reading code and documentation. (None of us is a Tcl expert.)

diego-ciciani01 committed:
fix: correct typos, update help text, and relocate IPv6 availability check

- Fixed minor typos across test files.

- Updated  to include the new option:

- Moved IPv6 availability check to tests/support/server.tcl.

Signed-off-by: diego-ciciani01 <diego.ciciani@gmail.com>

zuiderkwast commented on tests/support/server.tcl:
```diff
@@ -268,6 +268,11 @@ proc tags_acceptable {tags err_return} {
 
+    if {!$::require_ipv6 && ![is_ipv6_available]} {
+        lappend ::denytags "ipv6"
+        set err "IPv6 not available on this system, skipping IPv6 tests"
```
It appears you didn't get the purpose of this function and how it works.

This function runs for each `test` (and for `start_server` where tags can also be specified) and it checks it each test needs to be skipped or not.

For the command-line argument `--tags`, the `::denytags` and `::allowtags` are used. When a test is skipped because of `::denytags`, it prints "Tag: $tag denied". When a test is skipped for another reason, that specific reason is printed instead, such as "Not supported in cluster mode", etc. The reason I wanted you to move the check to this function is to have a separate check and separate message, and not hijack the command line `--tags` mechanism.

As you can see for all of the checks in this file, it sets `err` to an appropriate explanation and then returns 0.

```
    if {$::debug_defrag && [lsearch $tags "debug_defrag:skip"] >= 0} {
        set err "Not supported on server compiled with DEBUG_FORCE_DEFRAG option"
        return 0
    }

    if {$::singledb && [lsearch $tags "singledb:skip"] >= 0} {
        set err "Not supported on singledb"
        return 0
    }

    if {$::cluster_mode && [lsearch $tags "cluster:skip"] >= 0} {
        set err "Not supported in cluster mode"
        return 0
    }

    if {$::tls && [lsearch $tags "tls:skip"] >= 0} {
        set err "Not supported in tls mode"
        return 0
    }
```

Each check checks if `$tags` contains a specific tag and also checks if some condition is met, then sets `err` and returns 0 (which means skip the this test).

What you did here is to set `err` to "IPv6 not available on this system, skipping IPv6 tests" but the function still returns 1 below so the first test is not skipped, and as a side-effect, `ipv6` is added to `::denytags` as if `--tags -ipv6` was given on the command line. This means that the next test that has the `ipv6` will be skipped as if because `ipv6` is in `::denytags` and the error message will be "Tag: ipv6 denied". This is what is printed when `--tags -ipv6` is used on the command line. That's why I said "hijack" this machanism.

What the check needs to do instead is for check if "ipv6" is among the tags for this test and if it's not required and not available, then set the message and return 0. Something like...

```suggestion
    if {!$::require_ipv6 && [lsearch $tags "ipv6"] >= 0 && ![is_ipv6_available]} {
        set err "IPv6 not available on this system"
        return 0
```

We could also add an optimization to avoid running the auto-detect for every test case with the ipv6 tag (but if that's not too many tests, then it might be OK to run it every time).

diego-ciciani01 replied:
Now i get what you mean. For optimization of the test, we can save the result of IPv6 check on a variable, to avoid performing the test unnecessarily. I agree to take this into consideration when there are a considerable number of IPv6 tests. At the moment, I only count 5.

zuiderkwast commented on tests/test_helper.tcl:
```diff
@@ -687,6 +688,7 @@ proc print_help_screen {} {
         "--io-threads       Run tests with IO threads."
         "--tls              Run tests in TLS mode."
         "--tls-module       Run tests in TLS mode with Valkey module."
+        "--require-ipv6     Run tests forcing the ipv6."
```
I'm think if we need this flag or not. Comparing to other checks, I can see that we just skip tests that are not possible to run. For example we simply skip all ipv6 tests if we're running TCL 8.5. Similarly, some tests are skipped if TLS is disabled, if IO threads are used, and so on.

None of the other skipped tests have a `require` flag, so comparing to these other checks, I think we don't need this `--require-ipv6` flag. We can just skip them if it's not available.

Sorry for changing my mind! :smile:

diego-ciciani01 replied:
Don't worry, hahah.
 If I understand correctly, you just want to remove the print "--require-ipv6 Run tests forcing the ipv6.” 
 from, "print_help_screen" ? The meccanismo with the require_ipv6 variable remain the same ?

zuiderkwast replied:
> If I understand correctly, you just want to remove the print "--require-ipv6 Run tests forcing the ipv6.” from, "print_help_screen" ? The meccanismo with the require_ipv6 variable remain the same ?

No, I think we can remove the require_ipv6 variable completely.

There is no variables like that for other things that are only sometimes supported.

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -265,7 +265,7 @@ start_server {tags {"introspection"}} {
 
     start_server {tags {"ipv6"} overrides {bind {127.0.0.1 ::1}}} {
-        test {CLIENT LIST with IPv6 filter} {
+        test {CLIENT KILL with IPv6 filter} {
```
Good catch!

zuiderkwast commented on tests/unit/introspection.tcl:
```diff
@@ -518,7 +518,7 @@ start_server {tags {"introspection"}} {
 
             regexp {addr=\[([a-fA-F0-9:]+)\]:\d+} $client_info -> ipv6only
-            set filtered [$c client list not-ip "1.2.3.4" not-ip $ipv6only]
+            set filtered [$c client list not-ip $ipv6only]
```
This change looks unnecessary. Did you review your own diff after fixing the merge conflict?

diego-ciciani01 committed:
Fix is_ipv6_available() catch block behavior

- Is been removed the IPv6 flag

The return statement inside a catch block doesn't propagate out of
the procedure. Check catch return value (0 = success) instead.
Reference: https://www.tcl.tk/man/tcl8.6/TclCmd/catch.html

Signed-off-by: diego-ciciani01 <diego.ciciani@gmail.com>

diego-ciciani01 committed:
Fix typo in is_ipv6_available() comment

Signed-off-by: diego-ciciani01 <diego.ciciani@gmail.com>

zuiderkwast committed:
Undo trailing whitespace

Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

zuiderkwast committed:
Revert whitespace changes (try again)

Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

zuiderkwast commented:
> I hadn’t used cherry-pick, but I did regularly run git rebase upstream/unstable to keep my branch updated while the PR was open.

That's weird. I would think that should work. Maybe the problem is when you mix rebase and merge.

Thank you for your patience too!
